Auto prefiltering for queries on dense semantic_text fields

dimitris-athanasiou opened the pull request:
`knn` queries allow specifying `filters` that will be applied before the knn search. This `pre-filtering` allows the `knn` to return `k` results. If such filters are to applied only after the `knn` executes, then the `knn` returns the `k` matching results but the filters can filter out some of them thus potentially returning fewer than `k` results.

`semantic_text` fields can be queried with:

**DSL**

- `match` queries
- `semantic` queries
- `knn` queries

**ES|QL**

- `match` queries
- `knn` queries

For DSL, `knn` queries allow users to specify direct prefilters. However, `match` and `semantic` queries provide no way to do so. Same goes for ES|QL `MATCH`. Noting that ES|QL `KNN` already implements auto pre-filtering where conjunctions are pushed down to the `knn` query as prefilters.

This commit implements semantic_text auto pre-filtering for `semantic_text` queries in DSL (`match` and `semantic` queries) and ES|QL (`MATCH`).

We achieve this by adding an `AutoPrefilteringScope` object to the `SearchExecutionContext`. When we convert a `bool` query to a lucene query, we push its `must`, `filter`, and `must_not` clauses to the `AutoPrefilteringScope`. At that stage queries have already been rewritten. Semantic queries using `text_embedding` inference endpoints are rewritten to knn vector queries that are auto-prefiltering enabled. Then, when an auto-prefiltering enabled knn vector query is converted to its lucene equivalent, we fetch the prefilters from the `SearchExecutionContext` and we apply them to the knn vector query - which supports pre-filtering already.

ES|QL queries that contain `MATCH` automatically benefit from this implementation because they are rewritten in `bool` queries.

### Limitations

**DSL**

- nested queries are excluded from pre-filtering (#138184)

**ES|QL**

- filters that are not translatable to lucene queries will be applied as post-filters

Relates #132068

dimitris-athanasiou committed:
Auto prefiltering for queries on dense semantic_text fields

`knn` queries allow specifying `filters` that will be applied before
the knn search. This `pre-filtering` allows the `knn` to return `k`
results. If such filters are to applied only after the `knn` executes,
then the `knn` returns the `k` matching results but the filters can
filter out some of them thus potentially returning fewer than `k` results.

`semantic_text` fields can be queried with:

DSL

- `match` queries
- `semantic` queries
- `knn` queries

ES|QL

- `match` queries
- `knn` queries

For DSL, `knn` queries allow users to specify direct prefilters.
However, `match` and `semantic` queries provide no way to do so.
Same goes for ES|QL `match`. Noting that ES|QL `KNN` already implements
auto pre-filtering where conjunctions are pushed down to the `knn` query
as prefilters.

This commit implements semantic_text auto pre-filtering for `semantic_text` queries
in DSL (`match` and `semantic` queries) and ES|QL (`MATCH`).

We achieve this by adding an `AutoPrefilteringScope` object to the
`SearchExecutionContext`. When we convert a `bool` query to a lucene query,
we push its `must`, `filter`, and `must_not` clauses to the `AutoPrefilteringScope`.
At that stage queries have already been rewritten. Semantic queries using
`text_embedding` inference endpoints are rewritten to knn vector queries
that are auto-prefiltering enabled. Then, when an auto-prefiltering
enabled knn vector query is converted to its lucene equivalent, we
fetch the prefilters from the `SearchExecutionContext` and we apply
them to the knn vector query - which supports pre-filtering already.

ES|QL queries that contain `MATCH` automatically benefit from this implementation because
they are rewritten in `bool` queries.

Limitations

DSL

- nested queries are excluded from pre-filtering (#138184)

ES|QL

- filters that are not translatable to lucene queries will be applied as post-filters

Relates #132068

dimitris-athanasiou committed:
Update docs/changelog/138989.yaml

elasticsearchmachine commented:
Pinging @elastic/search-relevance (Team:Search - Relevance)

elasticsearchmachine commented:
Hi @dimitris-athanasiou, I've created a changelog YAML for you.

dimitris-athanasiou committed:
Changelog -> Vector Search

dimitris-athanasiou committed:
Sort by score in ES|QL test

dimitris-athanasiou committed:
Merge branch 'main' into semantic_text_auto_prefiltering

dimitris-athanasiou committed:
Make keyword field a filter clause in ES|QL test to exclude from scoring

dimitris-athanasiou committed:
Merge branch 'main' into semantic_text_auto_prefiltering

dimitris-athanasiou committed:
Merge branch 'main' into semantic_text_auto_prefiltering

dimitris-athanasiou committed:
Add better javadoc to AutoPrefilteringScope

dimitris-athanasiou committed:
Add a YAML test for semantic query

Mikep86 reviewed (commented):
Partial review, nice work! I didn't get to the tests, but I reviewed all the production logic. I identified some potential edge cases that I think we could iterate on.

dimitris-athanasiou commented on x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/SemanticMatchTestCase.java:
```diff
@@ -92,6 +92,64 @@ public void testCopyToWithSemanticText() throws IOException {
+            matchesList().item(matchesMap().entry("name", "keyword_field").entry("type", "keyword"))
+                .item(matchesMap().entry("name", "semantic_text_field").entry("type", "text")),
+            List.of(List.of("label_1", "violins"), List.of("label_1", "pianos"))
+        );
```
Unfortunately, all I get as profiler info is:

```
+(ConstantScore(#keyword_field:label_1 #single_value_match(keyword_field)))^0.0 +ToParentBlockJoinQuery (+ScoreAndDocQuery #_nested_path:semantic_text_field.inference.chunks)
```

regardless of whether I enable/disable auto-prefiltering. Thus, it seems that the profiler does not have a complete representation of the lucene query.

carlosdelest replied:
I _think_ you should be able to check prefilters in https://github.com/elastic/elasticsearch/blob/96e799acb87a90ba4522698da7b82bddb6a42bd9/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizerTests.java#L1783 - at the LocalPhysicalPlanOptimizer level you should have the query builders with the prefilters.

dimitris-athanasiou commented on server/src/main/java/org/elasticsearch/search/vectors/KnnVectorQueryBuilder.java:
```diff
@@ -626,6 +653,40 @@ protected Query doToQuery(SearchExecutionContext context) throws IOException {
+                Query query = queryBuilder.toQuery(context);
+                if (NestedHelper.containsNestedQuery(query) == false) {
+                    autoPrefilters.add(query);
+                }
```
We discussed this offline and then I just forgot about it 😞  You're right, we can do better here for `bool` queries and keep filters that aren't nested. I'll work on this.

dimitris-athanasiou commented on server/src/main/java/org/elasticsearch/search/vectors/KnnVectorQueryBuilder.java:
```diff
@@ -134,6 +138,18 @@ public static KnnVectorQueryBuilder fromXContent(XContentParser parser) {
+     * This is used to prevent infinite loops while applying auto pre-filtering as it is possible
+     * to have two knn queries where the one is a prefilter to the other.
+     */
+    private boolean isInTheMiddleOfAutoPrefiltering = false;
```
> If I understand correctly, this is to handle the scenario where we have multiple knn queries, and auto pre-filtering makes them pre-filters of each other. Is this correct?

Correct.

> If so, I think this could lead to some confusing behavior. For example, say we have knn queries A, B, and C, and they are processed in that order. Query A will get B and C as pre-filters. Query B will only get C as a pre-filter. Query C will have no pre-filters.

That is not how it works. Each knn query will get both other knn queries as prefilters.

> Changing the processing order could change the result set based on this opaque implementation detail. This doesn't seem ideal. Instead, what if we excluded all knn queries as pre-filters? We already know they make poor filters due to the limited value of k. We could implement this as a "filtered query set", and filter out nested queries in complex pre-filter queries using the same approach.

Based on my answer above, changing the order won't change the result. However, I do see how simply excluding `knn` queries from being used as prefilters might be the cleanest here. I'll see how this could be done.

dimitris-athanasiou commented on server/src/main/java/org/elasticsearch/search/vectors/KnnVectorQueryBuilder.java:
```diff
@@ -455,7 +482,7 @@ protected QueryBuilder doRewrite(QueryRewriteContext ctx) throws IOException {
                 rescoreVectorBuilder,
                 vectorSimilarity
-            ).boost(boost).queryName(queryName).addFilterQueries(filterQueries);
+            ).boost(boost).queryName(queryName).addFilterQueries(filterQueries).setAutoPrefilteringEnabled(isAutoPrefilteringEnabled);
```
To be fair the constructor situation in `KnnVectorQueryBuilder` has gone wild and I agree it is very fragile. A static helper would be nice but we'd still need to remember to call it. I wonder if the right solution here is a refactoring of the constructors. How about we leave this is follow up work? I can raise an issue for tidying this up.

dimitris-athanasiou committed:
Prune auto-prefilters

dimitris-athanasiou committed:
No need for loop protection while applying auto-prefiltering

dimitris-athanasiou committed:
Tests for pruning

dimitris-athanasiou committed:
Merge branch 'main' into semantic_text_auto_prefiltering

dimitris-athanasiou commented:
@Mikep86 I have pushed commits to the PR where I address your feedback:

- I have added a `AutoPrefilteringUtils.pruneQuery` that takes in a set of `QueryBuilder` classes, looks into the query tree and prunes query branches if from the query that matches one of the given classes. We only use `NestedQueryBuilder` for now.
- I have removed the loop protection in `KnnVectorQueryBuilder` as it is no longer necessary as we ensure queries exclude themselves from becoming prefilters in their inner queries.

dimitris-athanasiou committed:
Missing null check for minimum_should_match

dimitris-athanasiou committed:
Merge branch 'main' into semantic_text_auto_prefiltering

dimitris-athanasiou committed:
Merge branch 'main' into semantic_text_auto_prefiltering

Mikep86 commented on x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/100_semantic_text_auto_prefiltering.yml:
```diff
@@ -0,0 +1,936 @@
+
+  - match: { hits.total.value: 1 }
+  - match: { hits.hits.0._source.dense_field_1: "violins" }
+  - match: { hits.hits.0._source.dense_field_2: "cats" }
```
This is really more of a test showing where automatic pre-filtering doesn't work. Both `match` queries are rewritten to nested queries and excluded as pre-filters, which is why the "pianos" doc isn't returned. It would be nice to at least have a comment explaining that behavior and the expected results.

Mikep86 reviewed (commented):
Fantastic work 🙌 ! All the production code looks good. I pointed out a potential edge case in the min-should-match handing, but I don't have a good solution for it (it's also a very narrow edge case). Other than that, it's just a few small adjustments to tests.

dimitris-athanasiou committed:
Address review feedback on tests

dimitris-athanasiou committed:
Merge branch 'main' into semantic_text_auto_prefiltering

dimitris-athanasiou committed:
Better handling of minimum_should_match

dimitris-athanasiou committed:
Merge branch 'main' into semantic_text_auto_prefiltering

dimitris-athanasiou commented on server/src/main/java/org/elasticsearch/index/query/support/AutoPrefilteringUtils.java:
```diff
@@ -0,0 +1,134 @@
+            prunedBool.minimumShouldMatch(originalBool.minimumShouldMatch());
+        } else {
+            int originalMsm = Queries.calculateMinShouldMatch(originalBool.should().size(), originalBool.minimumShouldMatch());
+            prunedBool.minimumShouldMatch(originalMsm <= prunedBool.should().size() ? originalMsm : prunedBool.should().size());
```
Done in https://github.com/elastic/elasticsearch/pull/138989/commits/f4451e8d8d9820a8862592373bf59095753c0318

dimitris-athanasiou committed:
Merge branch 'main' into semantic_text_auto_prefiltering

dimitris-athanasiou committed:
Merge branch 'main' into semantic_text_auto_prefiltering

Mikep86 reviewed (approved):
LGTM 🚀

carlosdelest commented on server/src/main/java/org/elasticsearch/index/query/support/AutoPrefilteringUtils.java:
```diff
@@ -0,0 +1,149 @@
+            return Optional.empty();
+        }
+
+        if (query instanceof BoolQueryBuilder boolQuery) {
```
Nit - Consider using [pattern matching for switch ](https://docs.oracle.com/en/java/javase/17/language/pattern-matching-switch.html)

dimitris-athanasiou committed:
Merge branch 'main' into semantic_text_auto_prefiltering

carlosdelest reviewed (approved):
LGTM, amazing work!

It would be awesome to have a test in `SemanticMatchTestCase` that checks that ES|QL applies prefiltering - but not needed for this PR

dimitris-athanasiou commented on server/src/test/java/org/elasticsearch/search/vectors/AbstractKnnVectorQueryBuilderTestCase.java:
```diff
@@ -408,6 +412,43 @@ public void testMustRewrite() throws IOException {
     }
 
+    public void testBWCVersionSerialization_GivenAutoPrefiltering() throws IOException {
+        for (int i = 0; i < 100; i++) {
```
During my testing I found that I needed that to surface problems faster. It runs pretty fast so I left it in.

carlosdelest commented:
>> It would be awesome to have a test in SemanticMatchTestCase that checks that ES|QL applies prefiltering - but not needed for this PR

> @carlosdelest I have added such a test! It's there!

@dimitris-athanasiou  It indeed is! Isn't that awesome? 😅 🤦

dimitris-athanasiou committed:
Use pattern matching for switch

dimitris-athanasiou committed:
Use constant for number of test runs for BWC serialization

dimitris-athanasiou committed:
Merge branch 'main' into semantic_text_auto_prefiltering

dimitris-athanasiou committed:
Merge branch 'main' into semantic_text_auto_prefiltering
